packages/transport-tcp/src/listener.ts

```diff
@@ -205,69 +193,51 @@ export class TCPListener extends TypedEventEmitter<ListenerEvents> implements Li
 
-    try {
-      this.context.upgrader.upgradeInbound(maConn)
-        .then((conn) => {
-          this.log('inbound connection upgraded %s', maConn.remoteAddr)
-          this.connections.add(maConn)
-
-          socket.once('close', () => {
-            this.connections.delete(maConn)
-
-            if (
-              this.context.closeServerOnMaxConnections != null &&
-              this.connections.size < this.context.closeServerOnMaxConnections.listenBelow
-            ) {
-              // The most likely case of error is if the port taken by this
-              // application is bound by another process during the time the
-              // server if closed. In that case there's not much we can do.
-              // resume() will be called again every time a connection is
-              // dropped, which acts as an eventual retry mechanism.
-              // onListenError allows the consumer act on this.
-              this.resume().catch(e => {
-                this.log.error('error attempting to listen server once connection count under limit', e)
-                this.context.closeServerOnMaxConnections?.onListenError?.(e as Error)
-              })
-            }
-          })
+    this.context.upgrader.upgradeInbound(maConn)
+      .then((conn) => {
+        this.log('inbound connection upgraded %s', maConn.remoteAddr)
+        this.connections.add(maConn)
 
-          if (this.context.handler != null) {
-            this.context.handler(conn)
-          }
+        socket.once('close', () => {
+          this.connections.delete(maConn)
 
           if (
             this.context.closeServerOnMaxConnections != null &&
-            this.connections.size >= this.context.closeServerOnMaxConnections.closeAbove
+            this.connections.size < this.context.closeServerOnMaxConnections.listenBelow
           ) {
-            this.pause(false).catch(e => {
-              this.log.error('error attempting to close server once connection count over limit', e)
+            // The most likely case of error is if the port taken by this
+            // application is bound by another process during the time the
+            // server if closed. In that case there's not much we can do.
+            // resume() will be called again every time a connection is
+            // dropped, which acts as an eventual retry mechanism.
+            // onListenError allows the consumer act on this.
+            this.resume().catch(e => {
+              this.log.error('error attempting to listen server once connection count under limit', e)
+              this.context.closeServerOnMaxConnections?.onListenError?.(e as Error)
             })
           }
-
-          this.safeDispatchEvent('connection', { detail: conn })
         })
-        .catch(async err => {
-          this.log.error('inbound connection failed', err)
-          this.metrics?.errors.increment({ [`${this.addr} inbound_upgrade`]: true })
 
-          await attemptClose(maConn, {
```

Comment:
Was the source of the problem that `attemptClose` would close without aborting?

Reply:
Not exactly, it was calling `.close` which throw if closing gracefully fails, and then call `.abort`, but we were doing it in response to an error condition so we should just `.abort` and move on.

As it turns out this hasn't solved the problem completely but it has made it a little better so we're getting closer:


Before:

<img width="189" alt="image" src="https://github.com/user-attachments/assets/e6f41e04-756c-445a-a83a-0793f0cb6d61">

After:

<img width="235" alt="image" src="https://github.com/user-attachments/assets/ce5550ea-90ec-4203-8aee-1bfd76e2132f">

I think the bug may be in `@libp2p/tls`, without it the "inbound pending" count stays near 0 the whole time:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/aebf306e-2d13-4ad7-b5f8-2f45b924dd84">